src/src/PluginStructs/P145_data_struct.cpp

```diff
@@ -0,0 +1,655 @@
+@param[in] humidity     The relative humidity
+
+@return The corrected sensor resistance RZero in kOhm
+*/
+/*****************************************************************************/
+float P145_data_struct::getCorrectedRZero(float rSensor, float temperature, float humidity)
+{
+  float c;    // Correction factor
+
+  if (sensordef.cora <= 0.0)
+  {
+    return (getRZero(rSensor));
+  }
+  else if (temperature < 20)
+  {
+    c = sensordef.cora * temperature * temperature - sensordef.corb * temperature + sensordef.corc - (humidity - 33.) * sensordef.cord;
+  }
+  else
+  {
+    c = sensordef.core * temperature + sensordef.corf * humidity + sensordef.corg;
+  }
+  return (rSensor / c) * pow((refLevel / sensordef.para), (1. / sensordef.parb));
+}
+
+/*****************************************************************************/
+/*!
+@brief  Get the ppm of CO2 sensed (assuming only CO2 in the air)
+@param[in] sensorType   Sensor type identifier
+@param[in] rSensor      Sensor resistance value Rs
+@param[in] rZero        R0 Resistance zero value
+@return The ppm of CO2 in the air
+*/
+/*****************************************************************************/
+float P145_data_struct::getPPM(float rSensor)
+{
+  switch (algorithm)
+  {
+    case p145AlgA:  // MQ-135
+      return (sensordef.para * pow((rSensor/rzero), -sensordef.parb));
+      break;
+    case p145AlgB:  // Miquel5612 Exponential
+      return (sensordef.para *pow((rSensor/rzero), sensordef.parb));
+      break;
+    case p145AlgC:  // Miquel5612 Linear
+      return (pow(10, (log10(rSensor/rzero)-sensordef.parb)/sensordef.para));
+      break;
+    default:
+      return (0.0);
+      break;
+  }
+  return 0.0;   // Default value should never be returned
+}
+
+/*****************************************************************************/
+/*!
+@brief  Get the ppm of CO2 sensed (assuming only CO2 in the air), corrected
+        for temp/hum
+@param[in] rSensor      Sensor resistance value Rs
+@param[in] temperature  The ambient air temperature
+@param[in] humidity     The relative humidity
+@return The ppm of CO2 in the air
+*/
+/*****************************************************************************/
+float P145_data_struct::getCorrectedPPM(float rSensor, float temperature, float humidity)
+{
+  switch (algorithm)
+  {
+    case p145AlgA:
+      float c;                      // Temperature & humidity correction factor
+      if (temperature < 20)
+      {
+        c = sensordef.cora * temperature * temperature - sensordef.corb * temperature + sensordef.corc - (humidity - 33.) * sensordef.cord;
+      }
+      else
+      {
+        c = sensordef.core * temperature + sensordef.corf * humidity + sensordef.corg;
+      }
+      return (sensordef.para * pow(((rSensor / c) / rzero), -sensordef.parb));
+    case p145AlgB:
+      // TODO
+      break;
+    case p145AlgC:
+      // TODO
+      break;
+    default:
+      return (0.0);
+      break;
```

Comment:
No need for `break;` after a return.
And since you also return the same value at the end, you can leave out the return statement for the default case anyway.

Reply:
Defensive programming. Juggling between two thoughts 1) Return the intended velue for each case 2) Safety net at the end. As far as I know the compiler complains when not having a return at the end. I will clean it up a bit.